### PR TITLE
feat: Add fixed fee for creating a new credit class (#351)

### LINKED ISSUE #351: Add fixed fee for creating a new credit class
At first this should just involve sending fees to the fee pool, but eventually could change to send fees to some "credit class governance body" who stewards the larger set of on-chain credit classes.

- [ ] write specification
- [ ] decide about the pledge mechanism
- [ ] decide about the community pool

### PR DESCRIPTION
**Pull request by ruhatch**
This PR adds a fixed fee to the CreateClass method of `x/ecocredit`.

The fee is:
- charged to the `Designer` of the new class
- based on a global `x/params` parameter, updated through `x/gov`
- currently burned, but in future might be placed in a community pool, as per #383 

While reviewing this PR, please pay extra attention to:
- The module structure: Is this idiomatic Cosmos SDK module layout?
- The default genesis value:  Is this a sensible initial value?

Closes #351

### CODE REVIEW

File: x/ecocredit/module/module.go [hunk 1]
```diff
@@ -55,6 +68,36 @@ func (a Module) RegisterGRPCGatewayRoutes(sdkclient.Context, *runtime.ServeMux)
...
+
+// BeginBlock performs a no-op.
+func (Module) BeginBlock(_ sdk.Context, _ abci.RequestBeginBlock) {}
+
+// EndBlock returns the end blocker for the ecocredit module. It returns no
+// validator updates.
+func (Module) EndBlock(_ sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
+	return []abci.ValidatorUpdate{}
+}
+
```
Comment (blushi) [inline x/ecocredit/module/module.go:96]: I don't think those `AppModule` methods are needed here

File: x/ecocredit/module/module.go [hunk 2]
```diff
@@ -55,6 +68,36 @@ func (a Module) RegisterGRPCGatewayRoutes(sdkclient.Context, *runtime.ServeMux)
...
 /**** DEPRECATED ****/
 func (a Module) RegisterRESTRoutes(sdkclient.Context, *mux.Router) {}
 func (a Module) RegisterLegacyAminoCodec(*codec.LegacyAmino)       {}
+
+func (a Module) Route() sdk.Route                                    { return sdk.Route{} }
+func (a Module) QuerierRoute() string                                { return "" }
+func (a Module) LegacyQuerierHandler(*codec.LegacyAmino) sdk.Querier { return nil }
```
Comment (blushi) [inline x/ecocredit/module/module.go:103]: ditto

**Review by blushi — DISMISSED** [review]
Implementation looks good overall but I would rework the module wiring.

Could you resolve the conflicts?

File: third_party/proto/cosmos/base/query/v1beta1/pagination.proto [hunk 3]
```diff
@@ -30,6 +30,9 @@ message PageRequest {
   // count_total is only respected when offset is used. It is ignored when key
   // is set.
   bool count_total = 4;
+
+  // reverse is set to true if results are to be returned in the descending order.
+  bool reverse = 5;
```
Comment (aaronc) [inline third_party/proto/cosmos/base/query/v1beta1/pagination.proto:35]: This needs to be changed upstream in the SDK. We're just copying proto definitions here in `third_party`

Comment (aaronc) [reply third_party/proto/cosmos/base/query/v1beta1/pagination.proto:35]: nevermind, looks like this is just updating the files

Comment (ruhatch) [reply third_party/proto/cosmos/base/query/v1beta1/pagination.proto:35]: This came in from the `make proto-update-deps` command, so being copied from upstream

**Comment by aaronc** [comment]
Btw no need to rebase and force push @ruhatch. Since we're squashing everything you can just use merge commits

**Comment by ruhatch** [comment]
> Btw no need to rebase and force push @ruhatch. Since we're squashing everything you can just use merge commits

Nice, thanks for the tip!

File: x/ecocredit/server/server_test.go [hunk 4]
```diff
@@ -3,17 +3,62 @@ package server_test
...
+
+	// baseApp.Router().AddRoute(sdk.NewRoute(banktypes.ModuleName, bank.NewHandler(bankKeeper)))
+	baseApp.MountStore(authKey, sdk.StoreTypeIAVL)
+	baseApp.MountStore(bankKey, sdk.StoreTypeIAVL)
+	baseApp.MountStore(paramsKey, sdk.StoreTypeIAVL)
+	baseApp.MountStore(tkey, sdk.StoreTypeTransient)
+
+	authSubspace := paramstypes.NewSubspace(cdc, amino, paramsKey, tkey, authtypes.ModuleName)
+	bankSubspace := paramstypes.NewSubspace(cdc, amino, paramsKey, tkey, banktypes.ModuleName)
+	ecocreditSubspace := paramstypes.NewSubspace(cdc, amino, paramsKey, tkey, ecocredittypes.DefaultParamspace)
```
Comment (likhita-809) [inline x/ecocredit/server/server_test.go:44]:
```suggestion
	ecocreditSubspace := paramstypes.NewSubspace(cdc, amino, paramsKey, tkey, ecocredittypes.ModuleName)
```

File: x/ecocredit/server/server_test.go [hunk 5]
```diff
@@ -3,17 +3,62 @@ package server_test
...
+	amino := codec.NewLegacyAmino()
+
+	authtypes.RegisterInterfaces(cdc.InterfaceRegistry())
+
+	authKey := sdk.NewKVStoreKey(authtypes.StoreKey)
+	bankKey := sdk.NewKVStoreKey(banktypes.StoreKey)
+	paramsKey := sdk.NewKVStoreKey(paramstypes.StoreKey)
+	tkey := sdk.NewTransientStoreKey(paramstypes.TStoreKey)
+
+	// baseApp.Router().AddRoute(sdk.NewRoute(banktypes.ModuleName, bank.NewHandler(bankKeeper)))
```
Comment (likhita-809) [inline x/ecocredit/server/server_test.go:36]:
```suggestion
	baseApp.Router().AddRoute(sdk.NewRoute(banktypes.ModuleName, bank.NewHandler(bankKeeper)))
```

**Review by likhita-809 — COMMENTED** [review]
looks good, just few changes

File: x/ecocredit/server/msg_server.go [hunk 6]
```diff
@@ -37,6 +42,17 @@ func (s serverImpl) CreateClass(ctx types.Context, req *ecocredit.MsgCreateClass
...
+	// Charge the designer a fee to create the credit class
+	designerAddress, err := sdk.AccAddressFromBech32(req.Designer)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.ChargeCreditClassFee(ctx.Context, designerAddress)
+	if err != nil {
+		return nil, err
+	}
```
Comment (blushi) [inline x/ecocredit/server/msg_server.go:54]: shouldn't this be done before we actually create the credit class?

**Review by blushi — CHANGES_REQUESTED** [review]
I've tested it with the CLI and tx is failing with (coming from ABCI, recovering from a panic):
`panic message redacted to hide potentially sensitive system info: panic`
The tx is working fine on master. Can you reproduce it?

File: x/ecocredit/server/fee.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package server
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+)
+
+func (s serverImpl) GetCreditClassFee(ctx sdk.Context) sdk.Coins {
```
[the remark this answers is not in the document]
Comment (ruhatch) [reply x/ecocredit/server/fee.go:9]: Ah, yeah, it was in separate keeper package before, so couldn't be private

**Comment by ruhatch** [comment]
> I've tested it with the CLI and tx is failing with (coming from ABCI, recovering from a panic):
> `panic message redacted to hide potentially sensitive system info: panic`
> The tx is working fine on master. Can you reproduce it?

@blushi did you init a new chain or use an existing one? It might be panicking because it can't find the genesis parameters?

**Comment by blushi** [comment]
> > I've tested it with the CLI and tx is failing with (coming from ABCI, recovering from a panic):
> > `panic message redacted to hide potentially sensitive system info: panic`
> > The tx is working fine on master. Can you reproduce it?
> 
> @blushi did you init a new chain or use an existing one? It might be panicking because it can't find the genesis parameters?

Good point! That was indeed with an existing chain, I'll try out again with a new one.

**Review by blushi — APPROVED** [review]
tACK, nice work @ruhatch!

**Pull request merged**